txscript/engine.go

```diff
@@ -391,10 +397,10 @@ func (vm *Engine) CheckErrorCondition(finalScript bool) error {
 			"error check when script unfinished")
 	}
 
-	// If we're in witness execution mode, and this was the final script,
-	// then the stack MUST be clean in order to maintain compatibility with
-	// BIP16.
-	if finalScript && vm.witness && vm.dstack.Depth() != 1 {
+	// If we're in version zero witness execution mode, and this was the
+	// final script, then the stack MUST be clean in order to maintain
+	// compatibility with BIP16.
+	if finalScript && vm.isWitnessVersionActive(0) && vm.dstack.Depth() != 1 {
 		return scriptError(ErrEvalFalse, "witness program must "+
```

Comment:
Shouldn't this be `ErrCleanStack`?  I realize the reference tests expect `EVAL_FALSE`, but that seems incorrect to me.  The `parseExpectedResult` would just need to be updated to include `ErrCleanStack` as an acceptable error code for `EVAL_FALSE`.

Then, given that, it seems like it could be combined with the check below as:

```Go
if finalScript && (vm.hasFlag(ScriptVerifyCleanStack) || vm.isWitnessVersionActive(0)) && vm.dstack.Depth() != 1 {
```